### PR TITLE
Implement test-tracking in new tracing mode (IDEA-269908)

### PR DESCRIPTION
**Pull request by zuevmaxim**
https://youtrack.jetbrains.com/issue/IDEA-269908

- added test tracking tests
- `boolean[]` trace is saved into ClassData
- `traceLine` method is called in new tracing mode too, ClassData is saved into a field

### CODE REVIEW

File: src/com/intellij/rt/coverage/data/LineData.java [hunk 1]
```diff
@@ -41,7 +41,6 @@ public LineData(final int line, final String desc) {
 
   public void touch() {
     myHits++;
-    setTestName(ProjectData.getCurrentTestName());
```
Comment (akozlova) [inline src/com/intellij/rt/coverage/data/LineData.java:44]: won't old instrumenter break?

File: src/com/intellij/rt/coverage/data/ClassData.java [hunk 2]
```diff
@@ -34,6 +34,7 @@
 
   /** Storage for line and branch hits in new tracing mode. */
   private volatile int[] myHitsMask;
+  private volatile boolean[] myLinesTrace;
```
Comment (akozlova) [inline src/com/intellij/rt/coverage/data/ClassData.java:37]: please leave a comment what is it or give a better name; also can't we use only this array (of ints then) and on save append to the common hitsMask? Or won't it save some time?

File: src/com/intellij/rt/coverage/data/LineData.java (hunk 1 above)
Comment (zuevmaxim) [reply src/com/intellij/rt/coverage/data/LineData.java:44]: No, as I set test name in com.intellij.rt.coverage.data.ProjectData#testEnded.

Comment (zuevmaxim) [reply src/com/intellij/rt/coverage/data/LineData.java:44]: Moving test name set to the dumping stage reduces time from 10s to 7.3 in new tracing for joda.time. So, I'd make this test name set at the dumping stage. Also, I'll try to make trace saving on a separate thread.

Comment (zuevmaxim) [reply src/com/intellij/rt/coverage/data/LineData.java:44]: Parallel saving does not lead to a significant time decrease, so it was not implemented for now

**Review by akozlova — APPROVED** [review]

File: instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java [hunk 3]
```diff
@@ -38,9 +38,11 @@ public TouchCounter(MethodVisitor methodVisitor, BranchDataContainer branchData,
 
 
   public void visitLineNumber(int line, Label start) {
-    mv.visitVarInsn(Opcodes.ALOAD, getOrCreateLocalVariableIndex());
-    InstrumentationUtils.pushInt(mv, line);
-    mv.visitMethodInsn(Opcodes.INVOKESTATIC, ProjectData.PROJECT_DATA_OWNER, "trace", "(Ljava/lang/Object;I)V", false);
+    if (myBranchData.getContext().getLineData(line) != null) {
```
Comment (akozlova) [inline instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:41]: how is this related?

Comment (zuevmaxim) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:41]: It fixes test traking for tracing mode: ignored lines (closing brackets) should not be covered.

**Comment by zuevmaxim** [comment]
`idea.new.test.tracking.coverage` property was added to enable array mode of test tracking. Default value is true. Class data mode is used when this property is disabled.

**Comment by akozlova** [comment]
LGTM
